feat(api): Expand the supported resource pack operations for 1.20.3

zml2008 opened the pull request:
This PR makes it easy to handle multiple resource packs, and adds a callback system to track pack application.

(note, this does change some API used in snapshots so it'd be a binary break for any current testers)

zml2008 committed:
feat(api): Expand the supported resource pack operations for 1.20.3 features

zml2008 committed:
feat(api): Implement a helper to automatically compute resource pack hashes

zml2008 committed:
feat(api): Add a callback feature for resource pack application

Machine-Maker commented on api/src/main/java/net/kyori/adventure/resource/ResourcePackRequest.java:
```diff
@@ -191,6 +196,16 @@ interface Builder extends AbstractBuilder<ResourcePackRequest>, ResourcePackRequ
+     * @return a future providing the new resource pack request
+     * @since 4.15.0
+     */
+    @NotNull CompletableFuture<ResourcePackRequest> computeHashAndBuild();
```
Any use for providing your own executor optionally?

zml2008 replied:
Sure, that'd make sense I suppose.

Machine-Maker commented on api/src/main/java/net/kyori/adventure/audience/Audience.java:
```diff
@@ -708,24 +710,192 @@ default void openBook(final @NotNull Book book) {
+   */
+  @SuppressWarnings("checkstyle:MethodName")
+  @ForwardingAudienceOverrideNotRequired
+  default void setResourcePack(final @NotNull ResourcePackRequestLike request, final @NotNull ResourcePackRequestLike@NotNull... others) {
```
It does not make sense to accept more than one pack request for a "setResourcePack" method imo. If you want to have multiple, you use the send methods. set should just be a single one. Maybe making the name setResourcePacks would be better.

zml2008 replied:
The difference between set and send is, set is (clear + send), so why wouldn't it accept multiple. Making the method name plural would make sense I suppose.

Machine-Maker commented on api/src/main/java/net/kyori/adventure/audience/Audience.java:
```diff
@@ -708,24 +710,192 @@ default void openBook(final @NotNull Book book) {
+   * @see ResourcePackRequest
+   * @since 4.15.0
+   */
+  default void sendResourcePack(final @NotNull ResourcePackRequest request, final @NotNull ResourcePackRequest@NotNull... others) {
```
Can this method just delegate to the "send" methods with an empty callback? That way you don't have to implement multiple methods. You can make the empty callback some static constant that is available to implementations so they know they don't have to track it.

zml2008 replied:
eh, depending on the object identity of a no-op callback seems a bit iffy to me (unless we exposed it as API on `ResourcePackCallback` itself), but tbh as I think about it your suggestion makes more and more sense.

Machine-Maker reviewed (changes requested):
Should there be Iterable methods for sending/setting multiple as well?

zml2008 committed:
Support a user-provided executor for hash computation

zml2008 committed:
Pluralize pack sending methods

zml2008 committed:
introduce no-op callback concept to simplify implementations

zml2008 commented:
hm, looking at adding Iterable overloads, type erasure means we can't support all the individual types for each method. Right now, i feel like the best signature for implementations would be `thing(List<RPRequest>)`, where for API end users, the `thing(Iterable<? extends RPRequestLike>)` signature would make most sense -- but then we have to be careful about unboxing *Like instances pretty eagerly maybe?

zml2008 committed:
feat(api): Encapsulate resource pack request parameters

zml2008 committed:
feat(api): Introduce ResourcePackRequestLike

zml2008 committed:
how did this not trigger locally

zml2008 commented:
Looking at Vanilla again, it appears that while the client will accept multiple packs, it only uses the last `required` and `prompt` values sent, not one per pack. Maybe we want to push those values up to `Request` from `Info`?

Machine-Maker commented on api/src/main/java/net/kyori/adventure/resource/ResourcePackInfo.java:
```diff
@@ -0,0 +1,229 @@
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents a resource pack request that can be sent to players.
```
Don't think we should mention "request" in this type, now that info is separated from the request.

Machine-Maker commented on api/src/main/java/net/kyori/adventure/resource/ResourcePackInfo.java:
All, the mentions of "request" probably need to be handled.

kezz commented on api/src/main/java/net/kyori/adventure/audience/Audience.java:
```diff
@@ -708,24 +712,114 @@ default void openBook(final @NotNull Book book) {
-   * Sends a resource pack request to this audience.
+   * Sends a request to apply resource packs to this audience.
+   *
+   * <p>Multiple resource packs are only supported since 1.20.3. On older versions, all requests behave as if {@link ResourcePackRequest#replace()} is set to {@code true}.</p>
```
I wonder if we want to do something different here. Nobody is realistically going to want to set a bunch of packs with replace to true. I think we should just emulate this behaviour by only applying the final/first pack instead (effectively the same, but without the pointless network load/etc.

electronicboy replied:
I'd imagine that the only place in which replace would be true is for the first pack in the stack, "reset and apply these", assuming that the client supports that behavior (I would hope so)
as for legacy versions, I think that adventure should just recommend for a behavior here, i.e. impls should only care about grabbing and sending the last pack in the request?

assuming I'm not barking up the wrong tree for the discussion here

zml2008 replied:
`replace` applies to the overall request, not to individual packs within the request here -- that's the distinction between `ResourcePackInfo` and `ResourcePackRequest` really.

But yeah, the comments about legacy versions really only apply to whoever is masochistic enough to implement the resource pack API in -platform, modern impls can just ignore that bit.

zml2008 committed:
update pack info JD

zml2008 committed:
feat(api): Move some attributes up to the request level for resource packs

The `required` and `prompt` fields are only taken from the last pack
in the request in vanilla anyways, so they behave on what we call the
request level.

zml2008 committed:
add JD about the implications of required packs
